### ISSUE TITLE
Rendering hooks could change template rendering language

### ISSUE THREAD

**Issue by roadster31** [thelia/thelia#2065]
When a hook is rendered in a page, the language seems to be changed: 

Without hook rendering:

![img-2016-03-24_23-26-50](https://cloud.githubusercontent.com/assets/2197734/14033104/ea837bd4-f217-11e5-9424-b235375efbd8.png)

With hook rendering:

![img-2016-03-24_23-25-31](https://cloud.githubusercontent.com/assets/2197734/14033081/c1045526-f217-11e5-8109-4544cf269fb1.png)

**Comment by gillesbourgeat** [comment]
Hello,
It's the CustomerFamily module that changes the translation domain.
https://github.com/thelia-modules/CustomerFamily/blob/master/templates/frontOffice/default/account-update.html#L1

The problem would be same with an include.

**Comment by roadster31** [comment]
Yep, I've just realized this.

**Comment by gillesbourgeat** [comment]
@roadster31 yes :( I not know if there is a solution to avoid this behavior. 
The fact of always indicated the "d" argument  for fontion "intl" is also very annoying.

**Comment by roadster31** [comment]
> The fact of always indicated the "d" argument for fontion "intl" is also very annoying.
True.

Is there a way to detect if we're in a module from inside \TheliaSmarty\Template\Plugins\Translation ?

**Comment by gillesbourgeat** [comment]
For now, I have no solution.
I leave this issue opened.
Another contributor may have a solution.

**Comment by anoziere** [comment]
Ten years later, here is what happens, and a fix for the reported case.

The default translation domain is instance state of a singleton Smarty plugin (`TheliaSmarty/Template/Plugins/Translation.php:25`): `{default_translation_domain}` writes it (`:45-48`) and every `{intl}` reads it (`:76`). A hook listener renders its template through `BaseHook::render()` -> `SmartyParser::internalRenderer()`, that is a separate `fetch()` on the same Smarty instance and the same plugin instance, and nothing restores the previous value. So `{default_translation_domain domain='customerfamily.fo.default'}` at the top of CustomerFamily's `account-update.html` applies to everything rendered after the `{hook}` in the page: the theme strings are not in that catalog, the translator falls back to the source text, and the page looks untranslated. `{default_locale}` has exactly the same problem.

Three ways out:

1. **Save/restore around each hook render.** The `hook`/`hookblock` plugins snapshot the default domain and locale before dispatching the event and restore them afterwards (stacked, so nested hooks work, and in a `finally` so a failing listener cannot leave the domain changed). Local, additive, no behaviour change for modules: a hook template still inherits the ambient domain while it renders, it just cannot change it for the rest of the page.
2. **Scope the domain lexically, per template.** `{default_translation_domain}` would record the domain against the current Smarty template object and `{intl}` would resolve it by walking the parent chain. This is the only option that also fixes the `{include}` case @gillesbourgeat mentioned, but it depends on Smarty internals (inheritance vs include vs separate `fetch()`), and templates rendered by their own `fetch()` have the Smarty object as parent, so hook templates that today rely on the page domain (module templates overridden in the theme and translated in `fo.default`/`bo.default`) would suddenly resolve against an empty domain. That is a breaking change, not a 2.6 patch.
3. **Forbid modules to set the global domain / auto-detect the module.** Detecting "we are in a module" from the template path is wrong exactly where it matters (module templates overridden in the theme, theme templates rendered by a hook), and it would silently break every module that legitimately sets its domain.

I went for 1 in #3508, since it fixes the reported case with no BC impact and can be verified without a running shop (a fake listener that changes the domain no longer affects the caller). What it does *not* fix is `{include}` of a template that sets its own domain, and `{render}`/`{module_include}` used directly: those still need option 2, which would be a good candidate for a major version.

For the record, Thelia 3 is not affected: Smarty is gone and Twig's `trans_default_domain` is lexical, resolved at compile time for the template that declares it.

**Comment by anoziere** [comment]
Fixed in #3508, merged into the `2.6` branch: the `hook`/`hookblock` plugins now snapshot the default translation domain and locale before dispatching and restore them afterwards (stacked, `finally`-guarded), so a module template can no longer change the domain for the rest of the page. The `{include}` case discussed above still needs lexical per-template scoping, which is a major-version change (see the analysis in the previous comment); Thelia 3 is unaffected (Twig's `trans_default_domain` is lexical). Closing manually: `Fixes` keywords only auto-close from the default branch (`main`).
